scripts/remotes.lua

```
@@ -1,13 +1,217 @@
+--
+function shape_center(shape)
+  if shape.center then
+    return shape.center
+  else
+    return {x = 0.5 * (shape.left_top.x + shape.right_bottom.x), y = 0.5 * (shape.left_top.y + shape.right_bottom.y)}
+  end
+end
+
+--- Calculates the approximate center point between an array of circles
+--
+-- Calculation is a weighted average of the circle centers based on inverse radii weights (closer to smaller circles)
+--
+-- @param targets the array of circles
+--
+-- @return the approximate center point between the circles
+--
+function circle_centroid(targets)
+  local average = { x = 0.0, y = 0.0 }
+  local length = #targets
+  if length == 0 then
+    return average
+  end
+  -- approximate finding center point between circle edges
+  local weight = 0
+  for _, target in pairs(targets) do
+    if target.radius == nil then
+      return average -- quit early if we've not been passed all circles
+    end
+    -- limit modifier to radius 1 or greater to avoid issues with 0-radius inputs
+    local modifier = target.radius < 1.0 and (1/target.radius) or 2.0
+    weight = weight + modifier
+    average.x = average.x + modifier * target.center.x
+    average.y = average.y + modifier * target.center.y
+  end
+  average.x = average.x / weight
+  average.y = average.y / weight
+  return average
+end
+
+--- Calculates the approximate center point between an array of rectangles
+--
+-- Calculation is created by creating an average of each corner, then caulcating the center of the 'average' rectangle
+--
+-- @param targets the array of rectangles
+--
+-- @return the approximate center point between the rectangles
+--
+function rect_centroid(targets)
+  local avg_rect = {left_top={x=0.0,y=0.0},right_bottom={x=0.0,y=0.0}}
+  local length = #targets
+  if length == 0 then
+    return avg_rect
+  end
+  for _, rect in pairs(targets) do
+    if rect.left_top == nil then
+      return avg_rect -- quit early if we've not been passed all rectangles
```

If this is not meaningful, and is an error of passing in wrong arguments, better let the code crash so it does not mask errors in other parts of codebase.

---

Fair